### ISSUE TITLE
关于ASR与Clip tar测算

### ISSUE THREAD

**Issue by doudouwer** [gq-max/AdvDiffVLM#7]
您好！请问能否公开一下ASR与Clip tar测算的代码？

**Comment by doudouwer** [comment]
另外，我使用单张rtx titan运行main.py生成一张sample的时间仅仅为13秒，同时在llava上测试的效果非常差，想请问是否有超参数可以用来增加攻击效果？

**Comment by Zhou6661** [comment]
你是不是没有用迭代，我看论文好像要迭代多次

**Comment by doudouwer** [comment]
> 你是不是没有用迭代，我看论文好像要迭代多次

我的ddim_steps = 200，和论文中保持一致><还有其他地方需要配置的吗

**Comment by Zhou6661** [comment]
> > 你是不是没有用迭代，我看论文好像要迭代多次
> 
> 我的ddim_steps = 200，和论文中保持一致><还有其他地方需要配置的吗

论文里面有个超参数N，代码里面没有给，这个代码应该是个demo

**Comment by doudouwer** [comment]
> > > 你是不是没有用迭代，我看论文好像要迭代多次
> > 
> > 
> > 我的ddim_steps = 200，和论文中保持一致><还有其他地方需要配置的吗
> 
> 论文里面有个超参数N，代码里面没有给，这个代码应该是个demo

成功了，感谢大佬



```python
for i, ((image_org, _, path), (image_tgt, _, tgt_path)) in enumerate(zip(data_loader_imagenet, data_loader_target)):
    print(tgt_path)
    image_org = image_org.to(device)
    image_tgt = image_tgt.to(device)
    # # get tgt featutres
    with torch.no_grad():
        tgt_image_features_list=[]
        image_tgt = clip_preprocess(image_tgt)
        for clip_model in models:
            tgt_image_features = clip_model.encode_image(image_tgt)  # [bs, 512]
            tgt_image_features = tgt_image_features / tgt_image_features.norm(dim=1, keepdim=True)
            tgt_image_features_list.append(tgt_image_features)
            
    current_image = image_org.clone()

    for iteration in range(N_iteration):
        print('iteration:', iteration)
        with torch.no_grad():
            with model.ema_scope():
                uc = model.get_learned_conditioning(
                    {model.cond_stage_key: torch.tensor(n_samples_per_class*[1000]).to(model.device)}
                    )
                label_id = path[0].split('/')[-1].split('.')[0]
                #label_id = name_key[label_id]
                class_label = labeles[label_id]
                all_samples = list()
                all_labels = list()
                print(f"rendering {n_samples_per_class} examples of class '{class_label}' in {ddim_steps} steps and using s={scale:.2f}.")
                xc = torch.tensor(n_samples_per_class*[class_label])
                c = model.get_learned_conditioning({model.cond_stage_key: xc.to(model.device)})
                encoder_posterior = model.encode_first_stage(current_image)
                z = model.get_first_stage_encoding(encoder_posterior).detach()
                print(cam_root+label_id+'.png')
                cam = cv2.imread(cam_root+label_id+'.png', 0) / 255.
                cam = cv2.resize(cam, (64, 64))
                cam = torch.tensor(cam).float()
                samples_ddim, _ = sampler.sample(S=ddim_steps,
                                                conditioning=c,
                                                x_T=z,
                                                batch_size=n_samples_per_class,
                                                shape=[3, 64, 64],
                                                verbose=False,
                                                unconditional_guidance_scale=scale,
                                                unconditional_conditioning=uc,
                                                eta=ddim_eta,
                                                label=xc.to(model.device),
                                                tgt_image_features_list=tgt_image_features_list,
                                                org_image_features_list=None,
                                                cam=cam,
                                                K=1,s=35,a=5)


                x_samples_ddim = model.decode_first_stage(samples_ddim)
                if iteration == N_iteration - 1:
                    x_samples_ddim = torch.clamp((x_samples_ddim+1.0)/2.0, min=0.0, max=1.0)

        current_image = x_samples_ddim

        img_transformed = clip_preprocess(current_image).to(device) # image transformation to model input
        img_transformed_list.append(current_image)
        adv_image_feature_list = []
        for clip_model in models:
            adv_image_features = clip_model.encode_image(img_transformed)
            adv_image_features = adv_image_features / adv_image_features.norm(dim=1, keepdim=True)
            adv_image_feature_list.append(adv_image_features)

        #save_path = "./{:05d}_iter{:02d}.png".format(i, iteration)
        #torchvision.utils.save_image(final_preprocess(current_image), save_path)
    path = "./{:05d}.png".format(i)
    torchvision.utils.save_image(final_preprocess(current_image), path)
    break
```

**Comment by yikeqingli** [comment]
同求ASR测算代码

**Comment by ghost** [comment]
@doudouwer 您好，想请教一个问题。cle_data文件夹下的文件似乎要分为多个子文件夹，请问您是怎么做的？

**Comment by doudouwer** [comment]
> [@doudouwer](https://github.com/doudouwer) 您好，想请教一个问题。cle_data文件夹下的文件似乎要分为多个子文件夹，请问您是怎么做的？

起一个cle_images/sample文件夹即可，target_images也是这样。sample当作一个类

**Comment by ghost** [comment]
嗯，然后target_sample，他说是sd提前生成的，需要用到coco_Caption，请问你是如何得到target_sample文件的？
感谢

**Comment by doudouwer** [comment]
> 嗯，然后target_sample，他说是sd提前生成的，需要用到coco_Caption，请问你是如何得到target_sample文件的？ 感谢

emmm，就是下载个coco_caption的数据集，然后用stable_diffusion批量生成，放进sample文件夹就行

**Comment by ghost** [comment]
好的，谢谢您，我在理清一下哈。正常的流程应该是：
1、首先我从它给出的kaggle网站上把images和csv文件下载好，images文件夹也就是代码中的cle文件夹，然后在下面新建一个sample，把下载的数据都放进去。   
2、然后，我需要拿到cle_data/sample文件夹下对应的coco_caption，然后我用sd1.5，做text2image，然后放到target/sample下。
3、我用ipynb文件夹下提供的cam代码去生成可视化图片，然后放到cam_root下。
大概流程是这样吗？请问你那里还有保存的target文件夹的图片吗？或者对应的coco_caption，或者在哪里去找到它的现成的coco_caption呢？要不然我需要自己用打标工具打标，然后批量跑。
感谢

**Comment by doudouwer** [comment]
> 好的，谢谢您，我在理清一下哈。正常的流程应该是： 1、首先我从它给出的kaggle网站上把images和csv文件下载好，images文件夹也就是代码中的cle文件夹，然后在下面新建一个sample，把下载的数据都放进去。 2、然后，我需要拿到cle_data/sample文件夹下对应的coco_caption，然后我用sd1.5，做text2image，然后放到target/sample下。 3、我用ipynb文件夹下提供的cam代码去生成可视化图片，然后放到cam_root下。 大概流程是这样吗？请问你那里还有保存的target文件夹的图片吗？或者对应的coco_caption，或者在哪里去找到它的现成的coco_caption呢？要不然我需要自己用打标工具打标，然后批量跑。 感谢

你可以参考这个项目https://github.com/yunqing-me/AttackVLM的Prepare the scripts部分来生成。
流程是对的，gl

**Comment by ghost** [comment]
Thanks

**Comment by yikeqingli** [comment]
> > > > 你是不是没有用迭代，我看论文好像要迭代多次
> > > 
> > > 
> > > 我的ddim_steps = 200，和论文中保持一致><还有其他地方需要配置的吗
> > 
> > 
> > 论文里面有个超参数N，代码里面没有给，这个代码应该是个demo
> 
> 成功了，感谢大佬
> 
> for i, ((image_org, _, path), (image_tgt, _, tgt_path)) in enumerate(zip(data_loader_imagenet, data_loader_target)):
>     print(tgt_path)
>     image_org = image_org.to(device)
>     image_tgt = image_tgt.to(device)
>     # # get tgt featutres
>     with torch.no_grad():
>         tgt_image_features_list=[]
>         image_tgt = clip_preprocess(image_tgt)
>         for clip_model in models:
>             tgt_image_features = clip_model.encode_image(image_tgt)  # [bs, 512]
>             tgt_image_features = tgt_image_features / tgt_image_features.norm(dim=1, keepdim=True)
>             tgt_image_features_list.append(tgt_image_features)
>             
>     current_image = image_org.clone()
> 
>     for iteration in range(N_iteration):
>         print('iteration:', iteration)
>         with torch.no_grad():
>             with model.ema_scope():
>                 uc = model.get_learned_conditioning(
>                     {model.cond_stage_key: torch.tensor(n_samples_per_class*[1000]).to(model.device)}
>                     )
>                 label_id = path[0].split('/')[-1].split('.')[0]
>                 #label_id = name_key[label_id]
>                 class_label = labeles[label_id]
>                 all_samples = list()
>                 all_labels = list()
>                 print(f"rendering {n_samples_per_class} examples of class '{class_label}' in {ddim_steps} steps and using s={scale:.2f}.")
>                 xc = torch.tensor(n_samples_per_class*[class_label])
>                 c = model.get_learned_conditioning({model.cond_stage_key: xc.to(model.device)})
>                 encoder_posterior = model.encode_first_stage(current_image)
>                 z = model.get_first_stage_encoding(encoder_posterior).detach()
>                 print(cam_root+label_id+'.png')
>                 cam = cv2.imread(cam_root+label_id+'.png', 0) / 255.
>                 cam = cv2.resize(cam, (64, 64))
>                 cam = torch.tensor(cam).float()
>                 samples_ddim, _ = sampler.sample(S=ddim_steps,
>                                                 conditioning=c,
>                                                 x_T=z,
>                                                 batch_size=n_samples_per_class,
>                                                 shape=[3, 64, 64],
>                                                 verbose=False,
>                                                 unconditional_guidance_scale=scale,
>                                                 unconditional_conditioning=uc,
>                                                 eta=ddim_eta,
>                                                 label=xc.to(model.device),
>                                                 tgt_image_features_list=tgt_image_features_list,
>                                                 org_image_features_list=None,
>                                                 cam=cam,
>                                                 K=1,s=35,a=5)
> 
> 
>                 x_samples_ddim = model.decode_first_stage(samples_ddim)
>                 if iteration == N_iteration - 1:
>                     x_samples_ddim = torch.clamp((x_samples_ddim+1.0)/2.0, min=0.0, max=1.0)
> 
>         current_image = x_samples_ddim
> 
>         img_transformed = clip_preprocess(current_image).to(device) # image transformation to model input
>         img_transformed_list.append(current_image)
>         adv_image_feature_list = []
>         for clip_model in models:
>             adv_image_features = clip_model.encode_image(img_transformed)
>             adv_image_features = adv_image_features / adv_image_features.norm(dim=1, keepdim=True)
>             adv_image_feature_list.append(adv_image_features)
> 
>         #save_path = "./{:05d}_iter{:02d}.png".format(i, iteration)
>         #torchvision.utils.save_image(final_preprocess(current_image), save_path)
>     path = "./{:05d}.png".format(i)
>     torchvision.utils.save_image(final_preprocess(current_image), path)
>     break

ddim_mask里sample第一层循环的k应该就是N吧

**Comment by TheLiao233** [comment]
> 好的，谢谢您，我在理清一下哈。正常的流程应该是： 1、首先我从它给出的kaggle网站上把images和csv文件下载好，images文件夹也就是代码中的cle文件夹，然后在下面新建一个sample，把下载的数据都放进去。 2、然后，我需要拿到cle_data/sample文件夹下对应的coco_caption，然后我用sd1.5，做text2image，然后放到target/sample下。 3、我用ipynb文件夹下提供的cam代码去生成可视化图片，然后放到cam_root下。 大概流程是这样吗？请问你那里还有保存的target文件夹的图片吗？或者对应的coco_caption，或者在哪里去找到它的现成的coco_caption呢？要不然我需要自己用打标工具打标，然后批量跑。 感谢

你好，请问一下你已经能够正常运行并生成可以误导模型的对抗样本了吗？

**Comment by TheLiao233** [comment]
> > > > > 你是不是没有用迭代，我看论文好像要迭代多次
> > > > 
> > > > 
> > > > 我的ddim_steps = 200，和论文中保持一致><还有其他地方需要配置的吗
> > > 
> > > 
> > > 论文里面有个超参数N，代码里面没有给，这个代码应该是个demo
> > 
> > 
> > 成功了，感谢大佬
> > for i, ((image_org, _, path), (image_tgt, _, tgt_path)) in enumerate(zip(data_loader_imagenet, data_loader_target)):
> > print(tgt_path)
> > image_org = image_org.to(device)
> > image_tgt = image_tgt.to(device)
> > # # get tgt featutres
> > with torch.no_grad():
> > tgt_image_features_list=[]
> > image_tgt = clip_preprocess(image_tgt)
> > for clip_model in models:
> > tgt_image_features = clip_model.encode_image(image_tgt)  # [bs, 512]
> > tgt_image_features = tgt_image_features / tgt_image_features.norm(dim=1, keepdim=True)
> > tgt_image_features_list.append(tgt_image_features)
> > ```
> > current_image = image_org.clone()
> > 
> > for iteration in range(N_iteration):
> >     print('iteration:', iteration)
> >     with torch.no_grad():
> >         with model.ema_scope():
> >             uc = model.get_learned_conditioning(
> >                 {model.cond_stage_key: torch.tensor(n_samples_per_class*[1000]).to(model.device)}
> >                 )
> >             label_id = path[0].split('/')[-1].split('.')[0]
> >             #label_id = name_key[label_id]
> >             class_label = labeles[label_id]
> >             all_samples = list()
> >             all_labels = list()
> >             print(f"rendering {n_samples_per_class} examples of class '{class_label}' in {ddim_steps} steps and using s={scale:.2f}.")
> >             xc = torch.tensor(n_samples_per_class*[class_label])
> >             c = model.get_learned_conditioning({model.cond_stage_key: xc.to(model.device)})
> >             encoder_posterior = model.encode_first_stage(current_image)
> >             z = model.get_first_stage_encoding(encoder_posterior).detach()
> >             print(cam_root+label_id+'.png')
> >             cam = cv2.imread(cam_root+label_id+'.png', 0) / 255.
> >             cam = cv2.resize(cam, (64, 64))
> >             cam = torch.tensor(cam).float()
> >             samples_ddim, _ = sampler.sample(S=ddim_steps,
> >                                             conditioning=c,
> >                                             x_T=z,
> >                                             batch_size=n_samples_per_class,
> >                                             shape=[3, 64, 64],
> >                                             verbose=False,
> >                                             unconditional_guidance_scale=scale,
> >                                             unconditional_conditioning=uc,
> >                                             eta=ddim_eta,
> >                                             label=xc.to(model.device),
> >                                             tgt_image_features_list=tgt_image_features_list,
> >                                             org_image_features_list=None,
> >                                             cam=cam,
> >                                             K=1,s=35,a=5)
> > 
> > 
> >             x_samples_ddim = model.decode_first_stage(samples_ddim)
> >             if iteration == N_iteration - 1:
> >                 x_samples_ddim = torch.clamp((x_samples_ddim+1.0)/2.0, min=0.0, max=1.0)
> > 
> >     current_image = x_samples_ddim
> > 
> >     img_transformed = clip_preprocess(current_image).to(device) # image transformation to model input
> >     img_transformed_list.append(current_image)
> >     adv_image_feature_list = []
> >     for clip_model in models:
> >         adv_image_features = clip_model.encode_image(img_transformed)
> >         adv_image_features = adv_image_features / adv_image_features.norm(dim=1, keepdim=True)
> >         adv_image_feature_list.append(adv_image_features)
> > 
> >     #save_path = "./{:05d}_iter{:02d}.png".format(i, iteration)
> >     #torchvision.utils.save_image(final_preprocess(current_image), save_path)
> > path = "./{:05d}.png".format(i)
> > torchvision.utils.save_image(final_preprocess(current_image), path)
> > break
> > ```
> 
> ddim_mask里sample第一层循环的k应该就是N吧

您好，我把k的值改到20跑出来的图片放到llava去进行问答，模型回答的还是原图片的内容，请问您有生成能够误导模型的图片吗？

**Comment by gq-max** [comment]
ASR在少的时候可以人眼判断，多的时候可以使用GPT判断。原来的代码是一个demo，做了一个小更新

**Comment by Mogul-Ax** [comment]
> 好的，谢谢您，我在理清一下哈。正常的流程应该是： 1、首先我从它给出的kaggle网站上把images和csv文件下载好，images文件夹也就是代码中的cle文件夹，然后在下面新建一个sample，把下载的数据都放进去。 2、然后，我需要拿到cle_data/sample文件夹下对应的coco_caption，然后我用sd1.5，做text2image，然后放到target/sample下。 3、我用ipynb文件夹下提供的cam代码去生成可视化图片，然后放到cam_root下。 大概流程是这样吗？请问你那里还有保存的target文件夹的图片吗？或者对应的coco_caption，或者在哪里去找到它的现成的coco_caption呢？要不然我需要自己用打标工具打标，然后批量跑。 感谢

你好，请问你按照这个步骤成功浮现出论文了吗？方便加个v交流一下吗？
